mito-ai/mito_ai/providers.py

```diff
@@ -1,70 +1,81 @@
 import os
-from typing import AsyncGenerator, List, Optional, Union
+import traceback
+from typing import AsyncGenerator, Optional, Union
 
-from jinja2 import DictLoader, Environment
 from openai import AsyncOpenAI
-from openai.types.chat import ChatCompletionMessageParam, ChatCompletionChunk
 from openai._streaming import AsyncStream
-from traitlets import Unicode, default
+from openai.types.chat import ChatCompletionChunk
+from traitlets import CFloat, CInt, Unicode, default
 from traitlets.config import LoggingConfigurable
 
+from .logger import get_logger
 from .models import (
-    InlineCompletionItem,
-    InlineCompletionList,
-    InlineCompletionReply,
-    InlineCompletionRequest,
-    InlineCompletionStreamChunk,
+    CompletionError,
+    CompletionItem,
+    CompletionItemError,
+    CompletionReply,
+    CompletionRequest,
+    CompletionStreamChunk,
+)
+from .utils.db import get_user_field, set_user_field
+from .utils.open_ai_utils import get_ai_completion_from_mito_server
+from .utils.schema import UJ_AI_MITO_API_NUM_USAGES
+from .utils.telemetry_utils import (
+    KEY_TYPE_PARAM,
+    MITO_AI_COMPLETION_ERROR,
+    MITO_AI_COMPLETION_SUCCESS,
+    MITO_SERVER_KEY,
+    MITO_SERVER_NUM_USAGES,
+    USER_KEY,
+    log,
 )
 
 __all__ = ["OpenAIProvider"]
-
-
-COMPLETION_SYSTEM_PROMPT = """
-You are an application built to provide helpful code completion suggestions.
-You should only produce code. Keep comments to minimum, use the
-programming language comment syntax. Produce clean executable code.
-The code is written for a data analysis and code development
-environment which can execute code to produce graphics, tables and
-interactive outputs.
-"""
-
-
-COMPLETION_DEFAULT_TEMPLATE = """
-The document is called `{{filename}}` and written in {{language}}.
-"""
+_num_usages = None
 
 
 class OpenAIProvider(LoggingConfigurable):
     """Provide AI feature through OpenAI services."""
-    # Internally it uses jinja2 template to render prompt messages.
 
     api_key = Unicode(
         config=True,
         help="OpenAI API key. Default value is read from the OPENAI_API_KEY environment variable.",
     )
 
+    max_completion_tokens = CInt(
+        None,
+        allow_none=True,
+        config=True,
+        help="An upper bound for the number of tokens that can be generated for a completion, including visible output tokens and reasoning tokens.",
+    )
+
     # FIXME add validate function to check if the model is valid
     model = Unicode(
         "gpt-4o-mini", config=True, help="OpenAI model to use for completions"
     )
 
+    temperature = CFloat(
```

Comment:
Those are traitlets. So you can change them using the command line; e.g.

```sh
jupyter lab --OpenAIProvider.temperature=0.5
```

I did not yet add doc for this because I did not test them yet :wink: